src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,795 @@
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::trigger_out_of_safepoint_sampling() {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->trigger_out_of_safepoint_sampling();
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  _instance->handle_timer_signal(info, context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(siginfo_t* info, void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+  _sampler->handle_timer_signal(info, context);
+  Atomic::dec(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+}
+
+void JfrCPUTimeThreadSampler::sample_thread(JfrSampleRequest& request, void* ucontext, JavaThread* jt, JfrThreadLocal* tl) {
+  JfrSampleRequestBuilder::build_cpu_time_sample_request(request, ucontext, jt, jt->jfr_thread_local());
+}
+
+volatile size_t count__ = 0;
+
+static bool check_state(JavaThread* thread) {
+  switch (thread->thread_state()) {
+    case _thread_in_Java:
+    case _thread_in_native:
+      return true;
+    case _thread_in_vm:
+    default:
+      return false;
+  }
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(siginfo_t* info, void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrThreadLocal* tl = jt->jfr_thread_local();
+  if (!check_state(jt) ||
+      jt->is_at_poll_safepoint() ||
+      jt->is_JfrRecorder_thread()) {
+      tl->cpu_time_jfr_queue().increment_lost_samples();
+      tl->set_wants_out_of_safepoint_sampling(false);
+    return;
+  }
+  if (!tl->acquire_cpu_time_jfr_enqueue_lock()) {
+    tl->cpu_time_jfr_queue().increment_lost_samples();
+    return;
+  }
+
+  NoResourceMark rm;
+  JfrCPUTimeSampleRequest request;
+  // the sampling period might be too low for the current Linux configuration
+  // so samples might be skipped and we have to compute the actual period
+  int64_t period = get_sampling_period() * (info->si_overrun + 1);
+  request._cpu_time_period = Ticks(period / 1000000000.0 * JfrTime::frequency()) - Ticks(0);
+  sample_thread(request._request, context, jt, tl);
+
+  if (tl->cpu_time_jfr_queue().enqueue(request)) {
+    tl->set_has_cpu_time_jfr_requests(true);
+    SafepointMechanism::arm_local_poll_release(jt);
+  } else {
+    tl->cpu_time_jfr_queue().increment_lost_samples();
+  }
+
+  if (jt->thread_state() == _thread_in_native &&
+    tl->cpu_time_jfr_queue().size() > tl->cpu_time_jfr_queue().capacity() * 2 / 3) {
```

Comment:
Is this testing what you intend? The precedence order is [*, / ], from left to right associativity

Reply:
I'm sorry, I read this wrong. In this case, it does not matter: (capacity * 2) / 3 == capacity * (2/3)